### ISSUE TITLE
Custom CredentialProvider class not working correctly

### ISSUE THREAD

**Issue by nipun0505** [aws/aws-sdk-ios#561]
Hi, I tried to create my own credential provider class to add token to the flow, but it is not functioning correctly.

I get new token from my web API but when I send it to the AWS API, the error in the response says the token has been expired. The strange thing is, the error description has token string which I did not send. I sent the new one, but the error still shows the old token and labels it as expired.

Code:

```
#import "ABCAWSCredentialProvider.h"

@implementation ABCAWSCredentialProvider

+ (instancetype)credentialsWithAccessKey:(NSString *)accessKey
                               secretKey:(NSString *)secretKey sessionKey:(NSString*)sessionKey
{
    ABCAWSCredentialProvider *credentials = [[ABCAWSCredentialProvider alloc]initWithAccessKey:accessKey secretKey:secretKey sessionKey:sessionKey];
    return credentials;
    
}

- (instancetype)initWithAccessKey:(NSString *)accessKey
                        secretKey:(NSString *)secretKey sessionKey:(NSString*)sessionKey
{
    if (self = [super init]) {
        _accessKey = accessKey;
        _secretKey = secretKey;
        _sessionKey = sessionKey;
    }
    return self;
}


- (void)invalidateCachedTemporaryCredentials
{
    NSLog(@"Invalidate Called");
}

- (AWSTask<AWSCredentials *> *)credentials
{
    AWSCredentials *creds = [[AWSCredentials alloc] initWithAccessKey:_accessKey secretKey:_secretKey sessionKey:_sessionKey expiration:nil];
    
    return [AWSTask taskWithResult:creds];
}
@end
``` 

Using it like this:

```
 ABCAWSCredentialProvider *newProvider = [ABCAWSCredentialProvider credentialsWithAccessKey:access_KEY_ID secretKey:secret_KEY_ID sessionKey:sessionToken];

AWSServiceConfiguration *config=[[AWSServiceConfiguration alloc]initWithRegion:AWSRegionUSWest2 credentialsProvider:newProvider];
 
[AWSServiceManager defaultServiceManager].defaultServiceConfiguration = config;
```

**Comment by karthiksaligrama** [comment]
Are you using the developer authenticated identity flow? If you jsut have staticAWSCredentils then you could just use AWSStaticCredentialsProvider

**Comment by nipun0505** [comment]
I am note sure what you mean by "authenticated identity flow?". Could you please elaborate?

**Comment by karthiksaligrama** [comment]
resolved with 2.5.3. We added a new interface `AWSBasicSessionCredentialProvider` which would solve your use case
